File: 02-frontend/src/app/model/policy.ts
import {Client} from "./client";
import {Insurer} from "./insurer";
import {PolicyType} from "./policytype";
import {PaymentType} from "./paymenttype";
import {Agent} from "./agent";
import {User} from "./user";

export class Policy {
  public id: number;
  public client: Client;
  public insurer: Insurer;
  public policyType: PolicyType;
  public paymentType: PaymentType;
  public agent: Agent;
  public user: User;
  public conclusionDate: Date;
  public startDate: Date;
  public endDate: Date;
  public stopDate: Date;
  public reminderDate: Date;
  public reminder: boolean;
  public object: string;
  public folder: string;
  public policyNumber: string;
  public sum: number;
  public percent: number;
  public provision: number;
  public comment: string;
  public parts: number;

  constructor() {
    this.id = 0;
    this.client = null;
    this.insurer = null;
    this.policyType = null;
    this.paymentType = null;
    this.agent = null;
    this.user = null;
    this.conclusionDate = null;
    this.startDate = null;
    this.endDate = null;
    this.stopDate = null;
    this.reminderDate = null;
    this.reminder = true;
    this.object = '';
    this.folder = '';
    this.policyNumber = '';
    this.sum = 0;
    this.percent = 0;
    this.provision = 0;
    this.comment = '';
    this.parts = 0;
  }

  static fromJson(json: any): Policy {
    const policy = new Policy();

    policy.id = json.id;
    policy.client = json.client;
    policy.insurer = json.insurer;
    policy.policyType = json.policyType;
    policy.paymentType = json.paymentType;
    policy.agent = json.agent;
    policy.user = json.user;

    if (json.conclusionDate) policy.conclusionDate = new Date(json.conclusionDate);
    if (json.startDate) policy.startDate = new Date(json.startDate);
    if (json.endDate) policy.endDate = new Date(json.endDate);
    if (json.stopDate) policy.stopDate = new Date(json.stopDate);
    if (json.reminderDate) policy.reminderDate = new Date(json.reminderDate);

    policy.reminder = json.reminder;
    policy.object = json.object;
    policy.folder = json.folder;
    policy.policyNumber = json.policyNumber;
    policy.sum = json.sum;
    policy.percent = json.percent;
    policy.provision = json.provision;
    policy.comment = json.comment;
    policy.parts = json.parts;

    return policy;
  }

  static fromJsonList(jsonList: any[]): Policy[] {
    return jsonList.map(json => Policy.fromJson(json));
  }
}

File: 01-backend/src/main/java/ee/insa/crmapp/model/Policy.java
package ee.insa.crmapp.model;

import javax.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.LocalDate;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Entity
@Table(name = "policy")
public class Policy {
    @Id
    @Column(name = "id", nullable = false)
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @OneToOne
    private Client client;
    @OneToOne
    private Insurer insurer;
    @OneToOne
    private PolicyType policyType;
    @OneToOne
    private PaymentType paymentType;
    @OneToOne
    private Agent agent;
    @OneToOne
    private User user;
    private LocalDate conclusionDate;
    private LocalDate startDate;
    private LocalDate endDate;
    private LocalDate stopDate;
    private LocalDate reminderDate;
    private Boolean reminder;
    private String object;
    private String folder;
    private String policyNumber;
    private Double sum;
    private BigDecimal percent;
    private Double provision;
    @Column(length = 1000 )
    private String comment;
    private Integer parts;

    public Policy(Client client, Insurer insurer, PolicyType policyType, PaymentType paymentType, Agent agent, User user, LocalDate conclusionDate, LocalDate startDate, LocalDate endDate, LocalDate stopDate, LocalDate reminderDate, Boolean reminder, String object, String folder, String policyNumber, Double sum, BigDecimal percent, Double provision, String comment, Integer parts) {
        this.client = client;
        this.insurer = insurer;
        this.policyType = policyType;
        this.paymentType = paymentType;
        this.agent = agent;
        this.user = user;
        this.conclusionDate = conclusionDate;
        this.startDate = startDate;
        this.endDate = endDate;
        this.stopDate = stopDate;
        this.reminderDate = reminderDate;
        this.reminder = reminder;
        this.object = object;
        this.folder = folder;
        this.policyNumber = policyNumber;
        this.sum = sum;
        this.percent = percent;
        this.provision = provision;
        this.comment = comment;
        this.parts = parts;
    }

    @Override
    public String toString() {
        return "Policy{" +
                "id=" + id +
                ", client=" + client +
                ", insurer=" + insurer +
                ", policyType=" + policyType +
                ", paymentType=" + paymentType +
                ", agent=" + agent +
                ", user=" + user +
                ", conclusionDate=" + conclusionDate +
                ", startDate=" + startDate +
                ", endDate=" + endDate +
                ", stopDate=" + stopDate +
                ", reminderDate=" + reminderDate +
                ", reminder=" + reminder +
                ", object='" + object + '\'' +
                ", folder='" + folder + '\'' +
                ", policyNumber='" + policyNumber + '\'' +
                ", sum=" + sum +
                ", percent=" + percent +
                ", provision=" + provision +
                ", comment='" + comment + '\'' +
                ", parts=" + parts +
                '}';
    }
}
